nullaway/src/main/java/com/uber/nullaway/handlers/RxNullabilityPropagator.java

```diff
@@ -68,25 +72,69 @@
-    // should still be propagated to map(...), ignoring the interleaving call to m().
-    // We assume that if m() is a pass-through method for Observable, so are m(a1), m(a1,a2), etc. If that is ever not
-    // the case, we can use a more complex method subsignature her.
-    private final List<String> OBSERVABLE_PASSTHROUGH_METHOD_NAMES = ImmutableList.of(
-            "distinct",
-            "distinctUntilChanged",
-            "observeOn");
+    private final ImmutableList<StreamTypeRecord> RX_MODELS =
+            StreamModelBuilder.start()
+                .addStreamType(new DescendantOf(Suppliers.typeFromString("io.reactivex.Observable")))
+                    // Names of all the methods of io.reactivex.Observable that behave like .filter(...)
+                    // (must take exactly 1 argument)
+                    .withFilterMethodFromSignature("filter(io.reactivex.functions.Predicate<? super T>)")
+                    // Names and relevant arguments of all the methods of io.reactivex.Observable that behave like
+                    // .map(...) for the purposes of this checker (the listed arguments are those that take the
+                    // potentially filtered objects from the stream)
+                    .withMapMethodFromSignature(
+                            "<R>map(io.reactivex.functions.Function<? super T,? extends R>)",
+                            "apply",
+                            ImmutableSet.of(0))
+                    .withMapMethodAllFromName(
+                            "flatMap",
+                            "apply",
+                            ImmutableSet.of(0))
+                    .withMapMethodAllFromName(
+                            "flatMapSingle",
+                            "apply",
+                            ImmutableSet.of(0))
+                    .withMapMethodFromSignature(
+                            "distinctUntilChanged(io.reactivex.functions.BiPredicate<? super T,? super T>)",
+                            "test",
+                            ImmutableSet.of(0, 1))
+                    // List of methods of io.reactivex.Observable through which we just propagate the nullability
+                    // information of the last call, e.g. m() in Observable.filter(...).m().map(...) means the
+                    // nullability information from filter(...) should still be propagated to map(...), ignoring the
+                    // interleaving call to m().
+                    .withPassthroughMethodFromSignature("distinct()")
+                    .withPassthroughMethodFromSignature("distinctUntilChanged()")
+                    .withPassthroughMethodAllFromName("observeOn")
+                .addStreamType(new DescendantOf(Suppliers.typeFromString("io.reactivex.Maybe")))
+                    .withFilterMethodFromSignature("filter(io.reactivex.functions.Predicate<? super T>)")
```

Comment:
Can we avoid the code duplication between here and line 123?  Or are `Maybe` and `Single` different enough that it's worth keeping distinct copies of the code?

Reply:
So, the reason these models are separate is that the methods in those clases are completely separate as well (e.g. this in no inheritance relationship or common interface/supertype). Because of that, some methods are duplicated and others aren't and a few have similar names but different signatures. So it seems cleaner to me to keep them separate going forward, even if a few of these lines will end up being identical because of the similar APIs for these different stream objects.

Reply:
👍